versioned_docs/version-2021fa/lecture3.md

```diff
@@ -42,4 +42,364 @@ From now on, we will be using a `tsconfig.json` file within every Node project w
+  (i.e. there may be some delay in getting the response back)
+- SQL databases tend to be vertically scalable (need more computing power on one
+  machine to store more data) while NoSQL databases tend to be horizontally
+  scalable (can distribute storage and compute power on multiple machines)
+- Examples of SQL databases: MySQL, PostgreSQL
+- Examples of NoSQL databases: Fireabase, MongoDB
+
+### What is Firebase?
+
+- Firebase is a Backend as a Service (BaaS) offered by Google
+  - Allows you to store data
+  - Host websites
+  - Authentication
+- NoSQL DB
+  - Not only SQL
+  - Non relational
+
+### Why Firebase?
+
+- Real-time operations
+- Firebase Authentication
+- Built-in analytics
+- Also supports hosting/deployment
+- Integration with other Google services
+- Structure we’re familiar with!
+
+## Basic Database Manipulations
+
+People usually call that `CRUD`, which stands for:
+
+- **C**reate/Insert - Create a document (will fail if the document exists)
+- **R**ead/Find/Query - To search for documents based on their properties
+- **U**pdate - Update an existing document (will fail otherwise)
+- **D**elete - Delete an existing document
+
+For convenience, most NoSQL database also provides an _upsert_ operation. It
+will create the document or update an existing document. You can think of that
+as an atomic operation that does:
+
+```javascript
+if (document.exists()) {
+  database.update(document);
+} else {
+  database.insert(document);
+}
+```
+
+In Firestore, you can either insert a new document with a specified ID, or allow
+Firestore to generate its own ID for you.
+
+The update method in Firestore allows you to update certain fields of the
+document without overwriting the entire thing.
+
+## Sample code
+
+The following code demonstrates how we can do basic CRUD with Firestore. Note
+that the code below does not care about what are the fields of a post, because
+Firestore doesn't require you to have a predefined set of fields. This gives you
+flexibility when writing your backend code.
+
+```typescript title="index.ts"
+import admin from 'firebase-admin';
+import express from 'express';
+
+// require the service account: note the file path
+const serviceAccount = require('../service-account.json');
+// initialize the firebase app
+admin.initializeApp({
+  credential: admin.credential.cert(serviceAccount),
+});
+
+const db = admin.firestore();
+const app = express();
+const port = 8080;
+// allow request body parsing
+app.use(express.json());
+
+// check connections
+app.get('/', (_, res) => {
+  res.send('connected!');
+});
+
+// create a post type and post with id
+type Post = {
+  content: string;
+  name: string;
+};
+
+type PostWithID = Post & {
+  id: string;
+};
+
+// CRUD with firestore
+let posts1: Post[] = [{ content: 'I miss wellness days', name: 'Becky' }];
+
+// posts collection from db
+const postsCollection = db.collection('posts');
+
+// GET requests: get the songs
+app.get('/posts', (_, res) => {
```

Comment:
Can we rename these endpoint routes to be more clear about their functions?
Maybe like `/getPostsLocal` vs `/getPostsFirebase`